### PR TITLE
Implement readCursor

### PR DESCRIPTION
**Pull request by mheinzel**
# Description

Depends on #381. The code isn't pretty yet, but should work.

### CODE REVIEW

File: src/Database/LSMTree/Internal/Vector.hs [hunk 1]
```diff
@@ -95,3 +96,24 @@ unsafeInsertWithMStrict ::
...
+unfoldrNM' :: PrimMonad m => Int -> (b -> m (Maybe a, b)) -> b -> m (V.Vector a, b)
+unfoldrNM' len f = \b0 -> do
+    vec <- VM.unsafeNew len
+    go vec 0 b0
+  where
+    go vec !n !b
+      | n >= len = (, b) <$> V.unsafeFreeze vec
+      | otherwise =
+          f b >>= \case
+            (Nothing, b') ->
```
Comment (jorisdral) [inline src/Database/LSMTree/Internal/Vector.hs:115]:
```suggestion
            (Nothing, !b') ->
```

File: src/Database/LSMTree/Internal/Vector.hs [hunk 2]
```diff
@@ -95,3 +96,24 @@ unsafeInsertWithMStrict ::
...
+              (, b') <$> V.unsafeFreeze (VM.slice 0 n vec)
+            (Just a,  b') -> do
+              VM.unsafeWrite vec n a
```
Comment (jorisdral) [inline src/Database/LSMTree/Internal/Vector.hs:118]: Also, you might make this element strict, i.e. use `unsafeWrite vec n $! a`. This could also be reflected in the name -- and it might deserve a comment on just how strict the result actually is

File: src/Database/LSMTree/Internal/Vector.hs [hunk 3]
```diff
@@ -95,3 +96,24 @@ unsafeInsertWithMStrict ::
...
+-- /O(n)/ Construct a vector by repeatedly applying the monadic generator
+-- function to a seed. The generator function also yields 'Just' the next
+-- element or 'Nothing' if there are no more elements.
+unfoldrNM' :: PrimMonad m => Int -> (b -> m (Maybe a, b)) -> b -> m (V.Vector a, b)
+unfoldrNM' len f = \b0 -> do
+    vec <- VM.unsafeNew len
+    go vec 0 b0
+  where
+    go vec !n !b
+      | n >= len = (, b) <$> V.unsafeFreeze vec
```
Comment (jorisdral) [inline src/Database/LSMTree/Internal/Vector.hs:112]:
```suggestion
      | n >= len = (, b) <$!> V.unsafeFreeze vec
```

File: src/Database/LSMTree/Internal/Vector.hs [hunk 4]
```diff
@@ -95,3 +96,24 @@ unsafeInsertWithMStrict ::
...
+      | otherwise =
+          f b >>= \case
+            (Nothing, b') ->
+              (, b') <$> V.unsafeFreeze (VM.slice 0 n vec)
```
Comment (jorisdral) [inline src/Database/LSMTree/Internal/Vector.hs:116]:
```suggestion
              (, b') <$!> V.unsafeFreeze (VM.slice 0 n vec)
```

File: src/Database/LSMTree/Monoidal.hs [hunk 5]
```diff
@@ -306,7 +307,10 @@ readCursor ::
   => Int
   -> Cursor m k v
   -> m (V.Vector (QueryResult k v))
-readCursor = undefined
+readCursor n (Internal.MonoidalCursor c) =
+    Internal.readCursor const n c $ \k v mblob ->
+      assert (null mblob) $
```
Comment (jorisdral) [inline src/Database/LSMTree/Monoidal.hs:312]: This should probably be an `error` instead of an `assert`, like we do in `toMonoidalLookupResult`

File: src/Database/LSMTree/Internal.hs [hunk 6]
```diff
@@ -869,6 +877,112 @@ closeCursor Cursor {..} = do
         V.forM_ cursorRuns $ freeTemp reg . Run.removeReference
         return CursorClosed
 
+readCursor ::
+     forall m h res. m ~ IO -- TODO: replace by @io-classes@ constraints for IO simulation.
```
Comment (jorisdral) [inline src/Database/LSMTree/Internal.hs:881]: Needs a `SPECIALISE`

File: src/Database/LSMTree/Monoidal.hs [hunk 7]
```diff
@@ -306,7 +307,10 @@ readCursor ::
   => Int
   -> Cursor m k v
   -> m (V.Vector (QueryResult k v))
-readCursor = undefined
+readCursor n (Internal.MonoidalCursor c) =
+    Internal.readCursor const n c $ \k v mblob ->
```
Comment (jorisdral) [inline src/Database/LSMTree/Monoidal.hs:311]: This should probably not be `const`

File: test/Test/Database/LSMTree/Internal.hs [hunk 8]
```diff
@@ -207,13 +206,45 @@ prop_roundtripCursor kops = ioProperty $
...
     conf = testTableConfig
 
+    -- TODO: include inserts with blobs once write buffer returns BlobRefs
+    kopsWithoutblobs = V.map (fmap f) kops
+      where f (InsertWithBlob v _) = Insert v
+            f x                    = x
+
+    expected =
+      V.fromList . mapMaybe (traverse entryToValue) $
+          Map.assocs . Map.fromListWith (combine appendSerialisedValue) $
```
Comment (jorisdral) [inline test/Test/Database/LSMTree/Internal.hs:222]: Will this test ever encounter duplicates? It might be worth labelling so that we can see

File: src/Database/LSMTree/Internal.hs [hunk 9]
```diff
@@ -743,6 +748,8 @@ retrieveBlobs sesh wrefs =
   Cursors
 -------------------------------------------------------------------------------}
 
+-- TODO: Move to a separate Cursors module
```
Comment (jorisdral) [inline src/Database/LSMTree/Internal.hs:751]: :heavy_plus_sign:

**Review by jorisdral — COMMENTED** [review]
Not a full review yet, but here are some comments already

Let's also enable the statemachine tests

```diff
diff --git a/test/Test/Database/LSMTree/Normal/StateMachine.hs b/test/Test/Database/LSMTree/Normal/StateMachine.hs
index 0d3f4719..16311412 100644
--- a/test/Test/Database/LSMTree/Normal/StateMachine.hs
+++ b/test/Test/Database/LSMTree/Normal/StateMachine.hs
@@ -193,6 +193,7 @@ propLockstepIO_RealImpl_RealFS = testProperty "propLockstepIO_RealImpl_RealFS" $
       where
         handler' :: LSMTreeError -> Maybe Model.Err
         handler' ErrTableClosed = Just Model.ErrTableHandleClosed
+        handler' ErrCursorClosed = Just Model.ErrCursorClosed
         handler' (ErrSnapshotNotExists _snap) = Just Model.ErrSnapshotDoesNotExist
         handler' (ErrSnapshotExists _snap) = Just Model.ErrSnapshotExists
         handler' (ErrSnapshotWrongType _snap) = Just Model.ErrSnapshotWrongType
@@ -1033,8 +1034,7 @@ arbitraryActionWithVars _ findVars _st = QC.frequency $ concat [
       -> [(Int, Gen (Any (LockstepAction (ModelState h))))]
     withVars' genVar = [
           (2, fmap Some $ CloseCursor <$> (fromRight <$> genVar))
-          -- TODO: enable generators as we implement the actions for the /real/ lsm-tree
-        -- , (10, fmap Some $ ReadCursor <$> (QC.getNonNegative <$> QC.arbitrary) <*> (fromRight <$> genVar))
+        , (10, fmap Some $ ReadCursor <$> (QC.getNonNegative <$> QC.arbitrary) <*> (fromRight <$> genVar))
         ]

     withVars'' ::
```

It's probably also worth adding a few class-based tests for cursor, which can run against both the reference implementation and the real implementation

File: src/Database/LSMTree/Monoidal.hs (hunk 5 above)
Comment (mheinzel) [reply src/Database/LSMTree/Monoidal.hs:312]: Good catch!

File: src/Database/LSMTree/Monoidal.hs (hunk 7 above)
Comment (mheinzel) [reply src/Database/LSMTree/Monoidal.hs:311]: Good catch!

File: src/Database/LSMTree/Internal.hs (hunk 9 above)
Comment (mheinzel) [reply src/Database/LSMTree/Internal.hs:751]: I will do it in a follow-up PR.

**Comment by mheinzel** [comment]
I enabled the state machine tests and added class-based property tests. I also rebased and addressed some of the the feedback. The remainder will follow soon.

File: src/Database/LSMTree/Internal/Vector.hs (hunk 2 above)
Comment (mheinzel) [reply src/Database/LSMTree/Internal/Vector.hs:118]: I did it by adding a bang on the pattern match on `a` and added a comment.

File: src/Database/LSMTree/Monoidal.hs (hunk 5 above)
Comment (mheinzel) [reply src/Database/LSMTree/Monoidal.hs:312]: Not sure. It's an extra check for something that should never happen and we can handle it pretty gracefully. Similarly, we don't fail if there are mupserts when reading from a normal cursor, since the internal code just resolves it with the next entries, so it becomes indistinguishable from a single insert before being passed to the normal-specific code.

File: test/Test/Database/LSMTree/Internal.hs (hunk 8 above)
Comment (mheinzel) [reply test/Test/Database/LSMTree/Internal.hs:222]: I added labelling, or rather tabulation to also see how many keys in a test run are duplicates and how often they occur.

```
        +++ OK, passed 100 tests.
        
        duplicates (46 in total):
        85% 2
        13% 3
         2% 7
```

So not often, but they exist.

**Comment by mheinzel** [comment]
I think this is ready, apart from probably rebasing onto #346 later.

I also ran the benchmarks from #377 

```
benchmarking Bench.Database.LSMTree.Normal/cursor-scan-vs-range-lookup-scan/cursor-scan
time                 10.32 ms   (10.14 ms .. 10.52 ms)
                     0.996 R²   (0.992 R² .. 0.999 R²)
mean                 10.50 ms   (10.39 ms .. 10.71 ms)
std dev              396.5 μs   (256.9 μs .. 609.4 μs)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              2.903e7    (2.903e7 .. 2.903e7)
  y                  2620.153   (1709.444 .. 3490.414)
variance introduced by outliers: 16% (moderately inflated)
```

Which works out to reading about 7.75 million entries per second and 362 bytes/entry of allocations (keys and values are 8 byte each). Does that sound reasonable? Using `nfIO` instead of `whnfIO` doesn't change the result, so it seems like we are really forcing all the results.

File: test/Test/Database/LSMTree/Internal.hs [hunk 10]
```diff
@@ -207,13 +206,52 @@ prop_roundtripCursor kops = ioProperty $
...
+  -> IO (V.Vector (KeyForIndexCompact, (SerialisedValue, Maybe b)))
+readWholeCursor resolve cursor = go V.empty
+  where
+    chunkSize = 50
+    toResult k v b = (coerce k, (v, fmap toBlob b))
+
+    go !acc = do
+      res <- readCursor resolve chunkSize cursor toResult
+      if V.length res < chunkSize then return (acc <> res)
+                                  else go (acc <> res)
```
Comment (dcoutts) [inline test/Test/Database/LSMTree/Internal.hs:251]: Quadratic append, but probably doesn't matter for test case sizes. Otherwise accumulate chunks and reverse & concat at the end.

File: test/Test/Database/LSMTree/Normal/StateMachine.hs [hunk 11]
```diff
@@ -1033,8 +1034,7 @@ arbitraryActionWithVars _ findVars _st = QC.frequency $ concat [
       -> [(Int, Gen (Any (LockstepAction (ModelState h))))]
     withVars' genVar = [
           (2, fmap Some $ CloseCursor <$> (fromRight <$> genVar))
-          -- TODO: enable generators as we implement the actions for the /real/ lsm-tree
-        -- , (10, fmap Some $ ReadCursor <$> (QC.getNonNegative <$> QC.arbitrary) <*> (fromRight <$> genVar))
+        , (10, fmap Some $ ReadCursor <$> (QC.getNonNegative <$> QC.arbitrary) <*> (fromRight <$> genVar))
```
Comment (dcoutts) [inline test/Test/Database/LSMTree/Normal/StateMachine.hs:1037]: :tada:

**Review by dcoutts — APPROVED** [review]
Looks good!

File: test/Test/Database/LSMTree/Internal.hs (hunk 10 above)
Comment (mheinzel) [reply test/Test/Database/LSMTree/Internal.hs:251]: Yes, I just did the simplest thing here.

File: src/Database/LSMTree/Internal.hs [hunk 12]
```diff
@@ -869,6 +877,115 @@ closeCursor Cursor {..} = do
...
+                  Readers.Drained -> CursorOpen (cursorEnv {cursorReaders = Nothing})
+            return (state', vec)
+  where
+    -- General notes on the code below:
+    -- * it is quite similar to the one in Internal.Merge, but different enough
+    --   that it's probably easier to keep them separate
+    -- * any function that doesn't take a 'hasMore' argument assumes that the
+    --   readers have not been drained yet, so we must check before calling them
+    -- * there is probably opportunity for optimisations
+    readEntries hfs hbio readers =
```
[the remark this answers is not in the document]
Comment (mheinzel) [reply src/Database/LSMTree/Internal.hs:933]: Makes sense, I'll take your patch.

**Review by dcoutts — COMMENTED** [review]
LGTM!

**Comment by jorisdral** [comment]
> ...
> Which works out to reading about 7.75 million entries per second and 362 bytes/entry of allocations (keys and values are 8 byte each). Does that sound reasonable? Using `nfIO` instead of `whnfIO` doesn't change the result, so it seems like we are really forcing all the results.

For these specific benchmarks, there is no difference because the benchmarked action returns `IO ()`

**Comment by mheinzel** [comment]
> For these specific benchmarks, there is no difference because the benchmarked action has type `IO ()`

That's true, but when I tried this I changed `forM_` to `forM` as well.
